torch/csrc/distributed/c10d/PyProcessGroup.hpp

```diff
@@ -41,6 +41,19 @@ class PyProcessGroup : public ProcessGroup {
 
       return Work::getFuture();
     }
+
+    // Take a reference of the corresponding py::object.
+    // With functional collectives, ownership of work objects is generally
+    // transferred to C++. For pure C++ work objects, it is sufficient to
+    // transfer the ownership of work object. For user-defined work objects in
+    // Python, it is necessary to keep the corresponding py::object alive in
+    // addition to ensure that the user-defined methods can be executed.
+    void ref_py_object() {
+      py_obj_ = py::cast(this);
+    }
+
+   private:
+    py::object py_obj_;
```

Comment:
Work object lifetime is asynchronous with the life of the watchdog thread in C++.

Iirc we may not guarantee this thread exits before python shuts down. 

But if users call dist.destroy_process_group() as they are required to do, then the work objects should be finalized before python shuts down. 

Though, I don't think this PR does anything unsafe